README.txt
= ruby-rapleaf

* http://github.com/samstokes/ruby-rapleaf

== DESCRIPTION:

Rapleaf's goal is to make the internet safe and transparent, by allowing you
to evaluate your web footprint across a variety of user-generated sites.
See http://www.rapleaf.com for more information.

== FEATURES/PROBLEMS:

This library was inspired by the sample code provided by Rapleaf at:

http://trac.rapleaf.com/ruby-api-kit/wiki

The code was extensively re-written by Glenn Rempe so it would work with the
then-current Rapleaf v2 API, and further modified by Sam Stokes to support the
v3 API.

=== Original (v2 API compatible)

http://github.com/grempe/ruby-rapleaf

Glenn Rempe
glenn@rempe.us

=== Updated (v3 API compatible)

http://github.com/samstokes/ruby-rapleaf

Sam Stokes
sam@rapportive.com

== SYNOPSIS:

In order to use this API, you will need to get an API Key from Rapleaf. These
can be obtained free at http://www.rapleaf.com/.

To run the samples you will need to put your API Key in the
examples/person.rb file or the examples will fail.

To run any of the examples, simply run it like:

  cd examples
  ruby person.rb

(There are some known problems running the examples at the moment... sorry!)

== REQUIREMENTS:

The following gems are required for installation of the ruby-rapleaf gem.

* xml-simple
* builder

== INSTALL:

Sorry, there's no hosted gem at the moment!

You can build the gem yourself and install as a Ruby Gem with:

  git clone git://github.com/samstokes/ruby-rapleaf.git
  cd ruby-rapleaf
  rake gem
  rake install_gem

== LICENSE:

This software is distributed under the Ruby License.  A copy of which is
provided below.

RUBY LICENSE

Copyright (c) 2008-2010 Glenn Rempe, 2010 Sam Stokes

http://www.ruby-lang.org/en/LICENSE.txt

Ruby is copyrighted free software by Yukihiro Matsumoto <matz@netlab.co.jp>.
You can redistribute it and/or modify it under either the terms of the GPL
(see COPYING.txt file), or the conditions below:

  1. You may make and give away verbatim copies of the source form of the
     software without restriction, provided that you duplicate all of the
     original copyright notices and associated disclaimers.

  2. You may modify your copy of the software in any way, provided that
     you do at least ONE of the following:

       a) place your modifications in the Public Domain or otherwise
          make them Freely Available, such as by posting said
          modifications to Usenet or an equivalent medium, or by allowing
          the author to include your modifications in the software.

       b) use the modified software only within your corporation or
          organization.

       c) rename any non-standard executables so the names do not conflict
          with standard executables, which must also be provided.

       d) make other distribution arrangements with the author.

  3. You may distribute the software in object code or executable
     form, provided that you do at least ONE of the following:

       a) distribute the executables and library files of the software,
          together with instructions (in the manual page or equivalent)
          on where to get the original distribution.

       b) accompany the distribution with the machine-readable source of
          the software.

       c) give non-standard executables non-standard names, with
          instructions on where to get the original software distribution.

       d) make other distribution arrangements with the author.

  4. You may modify and include the part of the software into any other
     software (possibly commercial).  But some files in the distribution
     are not written by the author, so that they are not under this terms.

     They are gc.c(partly), utils.c(partly), regex.[ch], st.[ch] and some
     files under the ./missing directory.  See each file for the copying
     condition.

  5. The scripts and library files supplied as input to or produced as
     output from the software do not automatically fall under the
     copyright of the software, but belong to whomever generated them,
     and may be sold commercially, and may be aggregated with this
     software.

  6. THIS SOFTWARE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR
     IMPLIED WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED
     WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
     PURPOSE.


History.txt
=== 0.2.1 / 2010/04/27

* If unexpected response code, exception should include diagnostic info

=== 0.2.0 / 2010/04/24

* Support Person API lookup by site profile (e.g. Twitter username) for API v3.

=== 0.1.7 / 2010-04-14

* URLencode email addresses before sending them to Rapleaf.

* Remove workaround for mistakenly diagnosed problem (thought Rapleaf rejected emails containing '+', they just required encoding).

* Use existing PersonEmailHashNotFound exception for 404 when looking up by hash

  * removed NotFound class added in 0.1.6

=== 0.1.6 / 2010-03-04

* Modified by Sam Stokes

  * Support for Rapleaf Person API v3 (which is now the default)

  * Work around Rapleaf bug: they think email addresses containing '+' are malformed, so for those we send them hashes instead.

=== 0.1.5 / 2008-08-28

* Original version by Glenn Rempe, with support for Rapleaf Person API v2


lib/rapleaf/exceptions.rb
module Rapleaf

  # CLIENT SIDE ERRORS

  # All of our errors are superclassed by Error < RuntimeError
  class Error < RuntimeError #:nodoc:
  end

  # A client side only argument error
  class ArgumentError < Error #:nodoc:
  end

  # SERVER SIDE ERRORS

  # GENERAL API
  # =============================================

  # 401 Unauthorized
  # API key was not provided or is invalid.
  class AuthFailure < Error #:nodoc:
  end

  # 403 Forbidden
  # Your query limit has been exceeded.
  class ForbiddenQueryLimitExceeded < Error #:nodoc:
  end

  # 404 Not Found
  # Queried using a hash of an address Rapleaf didn't know.
  # (This doesn't happen when querying for the address directly, because
  # Rapleaf just creates a new person with that address if they didn't
  # have them already.)
  class NotFound < Error #:nodoc:
  end

  # 500 Internal Server Error
  # There was an unexpected error on our server. This should be very
  # rare and if you see it please contact developer@rapleaf.com.
  class InternalServerError < Error #:nodoc:
  end


  # PERSON API
  # =============================================

  # 202 Accepted
  # This person is currently being searched. Check back shortly and we should have data.
  class PersonAccepted < Error #:nodoc:
  end

  # 400 Bad Request
  # Invalid email address.
  class PersonBadRequestInvalidEmail < Error #:nodoc:
  end

  # 404 Not Found
  # Only returned for lookup by hash. We do not have this email in our
  # system and are not able to create a person using a hash. If you would
  # like better results, consider supplying the unhashed email address.
  class PersonEmailHashNotFound < Error #:nodoc:
  end


  # ABOOK API
  # =============================================

  # TODO

end


lib/rapleaf/rapleaf.rb
require 'md5'
require 'uri'

module Rapleaf

  class Base

    def initialize(api_key, options = {})
      options = {
                  :api_host     => API_HOST,
                  :api_port     => API_PORT,
                  :api_version  => API_VERSION
                }.merge(options)

      @api_key  = api_key
      @host     = options[:api_host]
      @port     = options[:api_port]
      @version  = options[:api_version]
    end

    # This resource is used to retrieve information about a person, identified
    # using an email address or email address hash.
    # Examples:
    #  person(:email => 'dummy@rapleaf.com')
    #  person(:site => :twitter, :profile => 'samstokes')
    #  person(:sha1 => SHA1.hexdigest('dummy@rapleaf.com'))
    #  person(:md5 => MD5.hexdigest('dummy@rapleaf.com'))
    def person( opts = {} )
      resp = Net::HTTP.get_response(URI.parse(person_url(opts)))

      case resp.code
      when '200'
        return Response.parse(:xml => resp.body)
      when '202'
        raise PersonAccepted, 'This person is currently being searched. Check back shortly and we should have data.'
      when '400'
        raise PersonBadRequestInvalidEmail, 'Invalid email address.'
      when '401'
        raise AuthFailure, 'API key was not provided or is invalid.'
      when '403'
        raise ForbiddenQueryLimitExceeded, 'Your query limit has been exceeded. Contact developer@rapleaf.com if you would like to increase your limit.'
      when '404'
        raise PersonEmailHashNotFound, 'We do not have this email in our system and are not able to create a person using a hash. If you would like better results, consider supplying the unhashed email address.'
      when '500'
        raise InternalServerError, 'There was an unexpected error on our server. This should be very rare and if you see it please contact developer@rapleaf.com.'
      else
        msg = resp.body[0,50]
        msg << "..." if 50 < resp.body.length
        raise Error, %(Unexpected response code #{resp.code}: "#{msg}")
      end
    end

  private
    def person_url(opts)
      email = opts[:email]

      site_profile = [opts[:site], opts[:profile]]
      if site_profile.any?
        raise ArgumentError, 'Require both :site and :profile if either is specified' unless site_profile.all?
      else
        site_profile = nil
      end

      md5 = opts[:md5]
      sha1 = opts[:sha1]

      # Rapleaf requires email addresses be urlencoded
      # Pass our own "unsafe regex" as URI.escape's default is too permissive
      # (lets + through, but Rapleaf rejects it)
      email = URI.escape(email, /[^a-zA-Z0-9.\-_]/) if email

      selector = [email, site_profile, md5, sha1].compact
      raise ArgumentError, 'Please provide only one of :email, [:site and :profile], :md5 or :sha1' if selector.size > 1
      raise ArgumentError, 'Person selector must be provided' if selector.empty? || '' == selector[0]

      case @version
      when "v2"
        if site_profile
          raise ArgumentError, 'Query by website ID requires API v3 or greater'
        end
        person_url_v2_by_email_or_hash(email_or_hash[0])
      when "v3"
        if email
          person_url_v3_by_email(email)
        elsif site_profile
          person_url_v3_by_site_profile(*site_profile)
        elsif md5
          person_url_v3_by_hash(:md5, md5)
        else
          person_url_v3_by_hash(:sha1, sha1)
        end
      else
        raise ArgumentError, "Person queries not supported for API version #{@version}"
      end
    end

    def person_url_v2_by_email_or_hash(email_or_hash)
      "http://#{@host}:#{@port}/v2/person/#{email_or_hash}?api_key=#{@api_key}"
    end

    def person_url_v3_by_email(email)
      "http://#{@host}:#{@port}/v3/person/email/#{email}?api_key=#{@api_key}"
    end

    def person_url_v3_by_site_profile(site, profile)
      # TODO validate param formats
      "http://#{@host}:#{@port}/v3/person/web/#{site}/#{profile}?api_key=#{@api_key}"
    end

    def person_url_v3_by_hash(algo, hash)
      "http://#{@host}:#{@port}/v3/person/hash/#{algo}/#{hash}?api_key=#{@api_key}"
    end

  end

end
